Compatibility with MSVC for the use of `restrict`

RoyiAvital opened the pull request:
Added a macro to support the use of `restrict` in MSVC.

Following on https://github.com/lemire/cbitset/issues/8.

lemire commented:
Defining a generic macro like RESTRICT won't work because it will be used in some other library.  One needs to do something like `CBITSET_RESTRICT`. Of course, it makes the whole code very ugly for the sole purpose of supporting Visual Studio which deliberately ignores the C standards.

A better solution might be...

```C
#if !defined(__clang__) && defined(_MSC_VER)
#ifndef restrict
#define restrict __restrict
#endif
#endif
```

It should only affect Visual Studio.

RoyiAvital commented:
I can do that.
Though, my preference would be the `CBITSET_RESTRICT`.
Though it will make the signatures longer, usually people will use those signatures as a library and not as their code.
So verbosity and clarity wins in my opinion.

Which one would you prefer?

I can also in the same time add the `calloc()` fix.

lemire commented:
> Though it will make the signatures longer, usually people will use those signatures as a library and not as their code.

I expect that people do read the source code, yes. 

> Which one would you prefer?

I am asking on twitter:
https://twitter.com/lemire/status/1308128061267816450

RoyiAvital commented:
> > Though it will make the signatures longer, usually people will use those signatures as a library and not as their code.
> 
> I expect that people do read the source code, yes.

I didn't mean they won't read it.
I meant they will see the extra letters in the signature only in the library and not in their code. Hence the _damage_ is low in my opinion.

RoyiAvital commented:
By the way, in order to reduce the added letters we can use `CBITSET_RSTRCT`.

Regarding the Twitter, I think `MSVC` allows variables with the name `restrict`.
Probably because they did not define it as reserved word because they use `__restrict`. So it takes the option off the table.

lemire commented:
@RoyiAvital Ok. Let us go with CBITSET_RESTRICT. Please don't use RSTRCT; it is terrible.

RoyiAvital commented:
@lemire , I updated the pull:

1. Using `CBIT_RESTRICT`.
2. Added case for any `C99` compatible compiler (I think the `restrict` word was defined in `C99`).
3. Using the `calloc()` function instead of `malloc()` + `memset(0)`.

Thank you for this library.

P. S.
There are some functions declared in the header yet not defined.
Are those stubs for future extension?

nemequ commented on include/portability.h:
```diff
@@ -2,6 +2,15 @@
+#define CBITSET_RESTRICT __restrict
+#else
+#define RESTRICT
+#endif // defined(__INTEL_COMPILER) || defined(__GNUC__) || defined(__GNUG__) || defined(__clang__) || (__STDC_VERSION__ >= 199901L)
```
Sorry for the drive-by pedantry from someone with no relationship to the project; feel free to ignore me!

I see a few problems with this if you care about portability.

 * This will result in errors from C++. `restrict` isn't supported, though `__restrict` is in most of them (in both C and C++ modes).  `restrict` should really be restricted to C99+ compilers w/o regard to the vendor
 * icl (ICC's MSVC-compatible frontend) won't work here as it doesn't support `restrict` but does define `__INTEL_COMPILER`.
 * `defined(__GNUG__)` doesn't do any good here; `__GNUC__` is defined in both C and C++ mode, `defined(__GNUG__)` is really just a shortcut for `defined(__GNUC__) && defined(__cplusplus)`

There is a [`HEDLEY_RESTRICT`](https://github.com/nemequ/hedley/blob/master/hedley.h#L1440-L1465) macro in Hedley which has some logic for this which you're welcome to steal (it's public domain). Hedley is one of my projects, though, so take the suggestion with a grain of salt :)

lemire replied:
Thanks Evan. I know of your work. This is a C library. Is hedley C compliant? If so, we could just add it to this project and be done with the whole issue.

nemequ replied:
Hedley is usable from C, yes.  Honestly, I'd say it's C first but also works in C++.

Technically I guess it's not really C "compliant"; it will take advantage of compiler-specific features (such as `__restrict`) when available, so it doesn't really preprocess to standards-compliant C when it knows the compiler can do better. But I don't think that's what you meant.

lemire commented:
Note that Visual Studio is adopting C11 :

https://devblogs.microsoft.com/cppblog/c11-and-c17-standard-support-arriving-in-msvc/

RoyiAvital commented:
@lemire , Good catch.
I think if `c99` and above is coming to MSVC I'd change the case to something much simpler.
Something like:

```
#if define(__STDC_VERSION__ >= 199901L) || (defined(__GNUC__) && defined(__STDC_VERSION__ ))
#define CBIT_RESTRICT restrict
#else 
#define CBIT_RESTRICT
#endif
```

Then it will be used only on `C` and the MSVC compatibility is here (For better performance on MSVC 19.8 people will have to use the `/std:c11` flag).

lemire commented:
@RoyiAvital This is indeed quite elegant.

lemire commented:
Merging.

nemequ commented:
That's definitely an improvement. Obviously this also wouldn't do any good in C++ mode, but the real problem is that `restrict` isn't actually accepted in C mode prior to C99 in either GCC or clang even as an extension. If you try to use this from C89 there will be [an error](https://godbolt.org/z/b58qTY).  Getting rid of the second condition would resolve this, and I don't think you would lose support for any compilers which would have otherwise worked.

I'm not sure how much you really care about this actually working in MSVC instead of just compiling (i.e., MSVC actually making use of `restrict`), but I would be wary about depending on the C11 support in MSVC… people tend not to upgrade expensive compilers like MSVC nearly as often as free compilers, so it will take a *long* time for C11 support which isn't even in a public release yet to reach every MSVC user. I haven't really asked around in a few years, but last time I checked I knew a *lot* of people who were still using 10+ year old versions of MSVC.

Anyways, it's already merged so I guess I'm a bit late.  Still, if nothing else I'd suggest dropping the ` || (defined(__GNUC__) && defined(__STDC_VERSION__))`

RoyiAvital commented:
@nemequ , Are you sure that by default most versions of `GCC` / `Clang` won't support `restrict` out of the box without even using `-std:c99`?

nemequ commented:
GCC currently defaults to C17, clang to C11.  The default changes from time to time… IIRC GCC has defaulted to C17 since gcc-7, clang to C11 since 3.6.

Most non-MSVC compilers are going to default to at least C99 these days, the problem is software which intentionally sticks to C89, which is pretty common. Since this is in a public header it's not like people can just compile cbitset in C99+ mode and their code in C89 mode; for that to work the cbitset header has to work in C89 mode.

lemire commented:
> Most non-MSVC compilers are going to default to at least C99 these days, the problem is software which intentionally sticks to C89, which is pretty common. Since this is in a public header it's not like people can just compile cbitset in C99+ mode and their code in C89 mode; for that to work the cbitset header has to work in C89 mode.

This library states openly that it requires C11. Support for Visual Studio is a bonus. No claim is made regarding older compilers.

nemequ commented:
> Support for Visual Studio is a bonus.

I'm extremely jealous.

lemire commented:
This is a C library. I take software seriously. If a vendor does not update its C compiler for 15 years, then they are signalling to you that they don’t take the language seriously.

Thankfully, it seems that Microsoft has changed its tune and they are planning to support C again.

You have to stand up against stupidity. Not supporting C99 in 2020 is stupid. Especially when you have more money than God. It is a clear sign that you don’t respect the language.

nemequ commented:
I completely agree; as a primarily C programmer Microsoft has spent the past couple decades disappointing and alienating me every chance they get, and I've desperately wanted to drop support for their compiler.  For most of my proprietary code I don't bother with MSVC support (or even Windows support), and it's wonderful.

Unfortunately abandoning MSVC support means abandoning a large pool of potential users.  Most people aren't going to switch to a better compiler so they can use my code; they're just not going to use my code.

This is further complicated by the fact that most people who want to use my projects want to do so from C++, and MSVC is a good C++ compiler (or so I'm told).  Trying to get someone to switch to a better compiler because MSVC's support for a language they don't really care about isn't going to work.

I really like that there are people like you who are willing to take a stand about this, but for me the downsides are too significant.

(Un?)fortunately I've accumulated quite a bit of code over the years to make this easier. Hedley is really helpful for abstracting away differences between C and C++ or using functionality which isn't available in all compilers without breaking compatibility. [Portable Snippets](https://github.com/nemequ/portable-snippets) has some handy stuff for things like calling `__builtin_*` functions from MSVC (see the "builtin" module), atomics, etc.  I could go on, but basically it's not usually *that* hard for me to support MSVC.

lemire commented:
> I really like that there are people like you who are willing to take a stand about this, but for me the downsides are too significant.

It depends on the context, but if I were to try to commercialize software written in C, I would give serious thought to using LLVM clang. It has full support for Windows and is fully integrated in Visual Studio 2019.

But even as far as libraries go... It really depends what the market is. There are always people who want support for GNU GCC 2.2. There are people who want Fortran support and COBOL. It is all fine, but are they willing to pay or are they hoping to get free lifetime support for their unpleasant systems ?


> Unfortunately abandoning MSVC support means abandoning a large pool of potential users. Most people aren't going to switch to a better compiler so they can use my code; they're just not going to use my code.

This is perfectly reasonable and I often choose to support Visual Studio if there is demand and people are willing to chip in.

> This is further complicated by the fact that most people who want to use my projects want to do so from C++, and MSVC is a good C++ compiler (or so I'm told). Trying to get someone to switch to a better compiler because MSVC's support for a language they don't really care about isn't going to work.

I would say that with the current Visual Studio (2019, with all the patches applied), Microsoft is probably ahead of the competition in C++ language support (though not necessarily in code generation).

It looks like Microsoft has decided to do the same with C (though they are only starting and have some work to do yet), so the point is less critical.

So it seriously look like Microsoft is offering a pleasant infrastructure for programmers and so I am increasingly willing to support them as a primary platform.  But if someone wants to run Windows 95 with Visual Studio 6, well... 

> (Un?)fortunately I've accumulated quite a bit of code over the years to make this easier. Hedley is really helpful for abstracting away differences between C and C++ or using functionality which isn't available in all compilers without breaking compatibility. Portable Snippets has some handy stuff for things like calling __builtin_* functions from MSVC (see the "builtin" module), atomics, etc. I could go on, but basically it's not usually that hard for me to support MSVC.

Yes. I know about Hedley as my friend Travis Downs uses it for everything. I was under the impression that it was C++ first which explains why I did not use it here. I will certainly consider it for future projects.
